### ISSUE TITLE
Java support

### ISSUE THREAD

**Issue by Groos-dev** [langgenius/dify-sandbox#31]
I can support the Java runtime environment

**Comment by Yeuoly** [comment]
Do you if you want : )

**Comment by Groos-dev** [comment]
yes！

**Comment by kanoshiou** [comment]
I am trying my best to support Java. I already have some ideas, but it might take some time.

**Comment by Groos-dev** [comment]
@kanoshiou I am supporting java. I am going to refer to the implementation method of nodejs, use jni to call the logic of java.so, and then execute the logic of the user input code. There are some issues that need to be verified. Does the limited system call using java.so have a restrictive effect on the execution of subsequent java programs? How to use third-party libraries in user input scripts? Maven? Or other

**Comment by Groos-dev** [comment]
Can I implement this feature? Could you please give me some suggestions.

**Comment by kanoshiou** [comment]
@Groos-dev Sure thing!


> use jni to call the logic of java.so

I would prefer using `jna` instead of `jni` for `jni` having strict rules for shared objects.

>Does the limited system call using java.so have a restrictive effect on the execution of subsequent java programs?

As far as I tested, the restriction works.

>How to use third-party libraries in user input scripts?

I think both Maven and Gradle are fine, but I feel that each has its shortcomings. For Maven, the XML file seems somewhat bloated; for Gradle, the jar file paths contain a layer of hashed folders, which might require some additional work.

**Comment by Groos-dev** [comment]
@kanoshiou 
Yes, JNA is more suitable for this scenario. Thank you very much for your suggestion. 
- I plan to spend two weeks implementing this function. The code repository address is https://github.com/Groos-dev/dify-sandbox/tree/feature/java-support. 
- In the use of package management tools, I prefer Maven because I am more familiar with Maven :). And I think that the cost and benefit of using Maven and Gradle in this scenario are not much different. If you think Gradle is more suitable, we can discuss it again.

**Comment by 18827555809** [comment]
> @kanoshiou Yes, JNA is more suitable for this scenario. Thank you very much for your suggestion.
> 
> * I plan to spend two weeks implementing this function. The code repository address is https://github.com/Groos-dev/dify-sandbox/tree/feature/java-support.
> * In the use of package management tools, I prefer Maven because I am more familiar with Maven :). And I think that the cost and benefit of using Maven and Gradle in this scenario are not much different. If you think Gradle is more suitable, we can discuss it again.

Hello, have there been any new developments recently? I noticed that the Java related support in the code repository is currently unavailable

**Comment by slieer** [comment]
Jshell integration should be feasible。

jutypter jshell
https://github.com/allen-ball/ganymede

**Comment by codeByJiaDong** [comment]
> * https://github.com/Groos-dev/dify-sandbox/tree/feature/java-support

你好，对java支持生效了吗
